Enhancment/artifacts documentation (#13724)
PrefectHQ/prefect: docs/3.0rc/concepts/artifacts.mdx

description: Prefect artifacts are persisted outputs designed for human consumption and available in the UI.
---

Artifacts are visually rich annotations on Prefect objects. They are visual metadata intended to be produced in flows or tasks but consumed by humans. They come in standardized formats such as tables, progress idicators, images, Markdown, or links. They are stored on Prefect Cloud or a Prefect server instance and rendered in the Prefect UI.
Artifacts make it easy to visualize outputs or side effects that your flows produce and capture updates over time.

![Markdown artifact sales report screenshot](/3.0rc/img/ui/md-artifact-info.png)

> Artifacts are associated with a particular Prefect object, usually a task run or flow run.

Reviewer: ```suggestion
Artifacts are associated with a particular Prefect object, usually a task run or a flow run.
```

- Data quality checks: Artifacts can be used to publish data quality checks from in-progress tasks. This can help ensure that data quality is maintained throughout the pipeline. During long-running tasks such as ML model training, you might use artifacts to publish performance graphs. This can help you visualize how well your models are performing and make adjustments as needed. You can also track the versions of these artifacts over time, making it easier to identify changes in your data.
- Documentation: Artifacts can be used to publish documentation and sample data to help you keep track of your work and share information with your colleagues. For instance, artifacts allow you to add a description to let your colleagues know why this piece of data is important.

## Creating artifacts

Creating artifacts allows you to publish data from task and flow runs or outside of a flow run context.
Currently, you can render five artifact types: links, Markdown, progress, images, and tables.

> !!! note "Artifacts render individually"

Reviewer: ```suggestion
<Note> 
```

>     Please note that every artifact created within a task will be displayed as an individual artifact in the Prefect UI.

Reviewer: ```suggestion
 Each artifact created within a task will be displayed individually in the Prefect UI. </Note>
```

description: Prefect artifacts are persisted outputs designed for human consumption and available in the UI.
---

> Artifacts are visually rich annotations on Prefect objects. They are visual metadata intended to be produced in flows or tasks but consumed by humans. They come in standardized formats such as tables, progress idicators, images, Markdown, or links. They are stored on Prefect Cloud or a Prefect server instance and rendered in the Prefect UI.

Reviewer 2: ```suggestion
Artifacts are rich annotations on Prefect runs. 
They are visual metadata intended to be produced in flow or task runs but consumed by humans. 
They come in standardized formats such as tables, progress indicators, images, Markdown, or links. 
They are stored on Prefect Cloud or a Prefect server instance and rendered in the Prefect UI.
```

> Artifacts make it easy to visualize outputs or side effects that your flows produce and capture updates over time.

Reviewer 2: ```suggestion
Artifacts make it easy to visualize outputs or side effects that your runs produce and capture updates over time.
```

![Markdown artifact sales report screenshot](/3.0rc/img/ui/md-artifact-info.png)

> Artifacts are associated with a particular Prefect object, usually a task run or a flow run.

Reviewer 2: Can they be associated with anything else?

Author: We don't want them to to global or unassociated anymore -- maybe there's a better way to say that?

Reviewer 2: I would just say "Artifacts are associated with a task or flow run."

description: Prefect artifacts are persisted outputs designed for human consumption and available in the UI.
---

> Artifacts are rich annotations on Prefect runs. 

Reviewer 2: ```suggestion
Artifacts are annotations on flow and task runs. 
```

> They are visual metadata intended to be produced in flow or task runs but consumed by humans. 

Reviewer 2: ```suggestion
They are visual metadata intended to be produced in code but consumed by humans. 
```

They come in standardized formats such as tables, progress indicators, images, Markdown, or links. 
They are stored on Prefect Cloud or a Prefect server instance and rendered in the Prefect UI.
Artifacts make it easy to visualize outputs or side effects that your runs produce and capture updates over time.

![Markdown artifact sales report screenshot](/3.0rc/img/ui/md-artifact-info.png)

> Artifacts are associated with a task run or a flow run.

Reviewer 2: ```suggestion
```

- Data quality checks: Artifacts can be used to publish data quality checks from in-progress tasks. This can help ensure that data quality is maintained throughout the pipeline. During long-running tasks such as ML model training, you might use artifacts to publish performance graphs. This can help you visualize how well your models are performing and make adjustments as needed. You can also track the versions of these artifacts over time, making it easier to identify changes in your data.
- Documentation: Artifacts can be used to publish documentation and sample data to help you keep track of your work and share information with your colleagues. For instance, artifacts allow you to add a description to let your colleagues know why this piece of data is important.

## Creating artifacts

Creating artifacts allows you to publish data from task and flow runs or outside of a flow run context.

> Currently, you can render five artifact types: links, Markdown, progress, images, and tables.

Reviewer 2: ```suggestion
There are five artifact types: links, Markdown, progress, images, and tables.
```

The `link` parameter is used to specify the external resource to be linked to, and `link_text` is used to specify the text to be displayed for the link.
An optional `description` could also be added for context.

### Creating progress artifacts

> Progress artifacts are useful for tracking the progress of long-running tasks. They render dynamically on the flow run timeline in the Prefect UI, providing a visually rich indicator of a task run's progress.

Reviewer 2: ```suggestion
Progress artifacts indicate the progress of long-running tasks. They render dynamically on the flow run graph in the Prefect UI, providing a visually rich indicator of a task run's progress.
```

To create a progress artifact, use the `create_progress_artifact()` function. To update the progress of a progress artifact, use the `update_progress_artifact()` function.

![Progress artifact example](/3.0rc/img/ui/progress-artifact-example.png)

```python
from time import sleep

from prefect import flow, task
from prefect.artifacts import (
    create_progress_artifact,
    update_progress_artifact,
)

def fetch_batch(i: int):
    # Simulate fetching a batch of data
    sleep(2)

@task
def fetch_in_batches():
    progress_artifact_id = create_progress_artifact(
        progress=0.0,
        description="Indicates the progress of fetching data in batches.",
    )
    for i in range(1, 11):
        fetch_batch(i)
        update_progress_artifact(artifact_id=progress_artifact_id, progress=i * 10)

@flow
def etl():
    fetch_in_batches()

if __name__ == "__main__":
    etl()

```

> Note that progress artifacts are updated using a separate `update_progress_artifact()` function. Prefect updates a progress artifact in place, rather than versioning it.

Reviewer 2: ```suggestion
Progress artifacts are updated using a separate `update_progress_artifact()` function. Prefect updates a progress artifact in place, rather than versioning it.
```

> Note that progress artifacts are updated using a separate `update_progress_artifact()` function. This is because we're not tracking revisions between versions of an artifact, we're updating an existing artifact in place.

Reviewer: ```suggestion
Note that progress artifacts are updated using a separate `update_progress_artifact()` function. Prefect updates a progress artifact in place, rather than versioning it.
```

The `link` parameter is used to specify the external resource to be linked to, and `link_text` is used to specify the text to be displayed for the link.
An optional `description` could also be added for context.

### Creating progress artifacts

> Progress artifacts indicate the progress of long-running tasks. They render dynamically on the flow run graph in the Prefect UI, providing a visually rich indicator of a task run's progress.

Reviewer 2: ```suggestion
Progress artifacts render dynamically on the flow run graph in the Prefect UI, indicating the progress of long-running tasks. 
```

> To create a progress artifact, use the `create_progress_artifact()` function. To update the progress of a progress artifact, use the `update_progress_artifact()` function.

Reviewer 2: ```suggestion
To create a progress artifact, use the `create_progress_artifact()` function. To update a progress artifact, use the `update_progress_artifact()` function.
```

As you can see, you don't need to create an artifact in a flow run context.
You can create one anywhere in a Python script and see it in the Prefect UI.

### Create image artifacts

> Image artifacts render publically available images in the Prefect UI for easy visualization & association with a task or flow run. To create an image artifact, use the `create_image_artifact()` function.

Reviewer 2: ```suggestion
Image artifacts render publicly available images in the Prefect UI. To create an image artifact, use the `create_image_artifact()` function.
```

![Image artifact example](/3.0rc/img/ui/image-artifact-example.png)

```python
from prefect import flow, task
from prefect.artifacts import (
    create_image_artifact,
)

@task
def create_image():
    # Do something to create an image and upload to a url
    image_url = "https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExZmQydzBjOHQ2M3BhdWJ4M3V1MGtoZGxuNmloeGh6b2dvaHhpaHg0eSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3KC2jD2QcBOSc/giphy.gif"
    create_image_artifact(image_url=image_url, description="A gif.", key="gif")
    return image_url

@flow
def my_flow():
    return create_image()

if __name__ == "__main__":
    image_url = my_flow()
    print(f"Image URL: {image_url}")

```

> Note: The image URL must be publicly accessible for the Prefect UI to render it. If you would like to create an artifact for a private image, you can use the `create_link_artifact()` function instead.

Reviewer 2: ```suggestion
To create an artifact that links to a private image, use the `create_link_artifact()` function instead.
```
